File: src/entropy.ts
// Entropy and Information Theory utilities for Wordle
// Based on information theory principles where entropy measures expected information gain

export type LetterResult = 'B' | 'Y' | 'G'; // Black, Yellow, Green
export type Pattern = string; // e.g., "BGYBB"

/**
 * Generates the pattern (response) when guessing a word against a target word
 * B = Black (letter not in word)
 * Y = Yellow (letter in word, wrong position) 
 * G = Green (letter in word, correct position)
 */
export function getPattern(guess: string, target: string): Pattern {
    const guessLetters = guess.toUpperCase().split('');
    const targetLetters = target.toUpperCase().split('');
    const result: LetterResult[] = new Array(5).fill('B');

    // Create a frequency map of target letters for yellow/black logic
    const targetFreq: { [key: string]: number } = {};
    for (const letter of targetLetters) {
        targetFreq[letter] = (targetFreq[letter] || 0) + 1;
    }

    // First pass: mark green letters and reduce frequency
    for (let i = 0; i < 5; i++) {
        if (guessLetters[i] === targetLetters[i]) {
            result[i] = 'G';
            targetFreq[guessLetters[i]]--;
        }
    }

    // Second pass: mark yellow letters
    for (let i = 0; i < 5; i++) {
        if (result[i] === 'B' && targetFreq[guessLetters[i]] > 0) {
            result[i] = 'Y';
            targetFreq[guessLetters[i]]--;
        }
    }

    return result.join('');
}

/**
 * Checks if a word matches the given constraints
 */
export function matchesConstraints(
    word: string,
    knownPositions: string[],
    yellowLetters: Array<{ letter: string, excludedPositions: number[] }>,
    grayLetters: string[]
): boolean {
    const upperWord = word.toUpperCase();

    // Check known positions (green constraints)
    for (let i = 0; i < knownPositions.length; i++) {
        if (knownPositions[i] && upperWord[i] !== knownPositions[i]) {
            return false;
        }
    }

    // Check yellow letters (must be in word but not in excluded positions)
    for (const yellowInfo of yellowLetters) {
        const letter = yellowInfo.letter;

        // Letter must be present in the word
        if (!upperWord.includes(letter)) {
            return false;
        }

        // Letter must not be in any of the excluded positions
        for (const excludedPos of yellowInfo.excludedPositions) {
            if (excludedPos < upperWord.length && upperWord[excludedPos] === letter) {
                return false;
            }
        }
    }

    // Check gray letters (must not be in word)
    for (const letter of grayLetters) {
        if (upperWord.includes(letter)) {
            return false;
        }
    }

    return true;
}

/**
 * Calculates the entropy (expected information gain) for a guess word
 * Higher entropy = better guess (more information gained on average)
 */
export function calculateEntropy(guessWord: string, possibleAnswers: string[]): number {
    if (possibleAnswers.length <= 1) {
        return 0;
    }

    // Count frequency of each pattern that could result from this guess
    const patternCounts: { [pattern: string]: number } = {};

    for (const answer of possibleAnswers) {
        const pattern = getPattern(guessWord, answer);
        patternCounts[pattern] = (patternCounts[pattern] || 0) + 1;
    }

    // Calculate entropy using Shannon's formula: H = -Σ p(x) * log₂(p(x))
    let entropy = 0;
    const totalAnswers = possibleAnswers.length;

    for (const count of Object.values(patternCounts)) {
        const probability = count / totalAnswers;
        entropy -= probability * Math.log2(probability);
    }

    return entropy;
}

/**
 * Calculates entropy for all words and returns them sorted by entropy (highest first)
 */
export function calculateAllEntropies(
    guessWords: string[],
    possibleAnswers: string[]
): Array<{ word: string; entropy: number; bitsOfInfo: number }> {
    const results = guessWords.map(word => {
        const entropy = calculateEntropy(word, possibleAnswers);
        return {
            word: word.toUpperCase(),
            entropy,
            bitsOfInfo: Number(entropy.toFixed(2))
        };
    });

    // Sort by entropy (highest first)
    return results.sort((a, b) => b.entropy - a.entropy);
}

/**
 * Gets the top N words with highest entropy
 */
export function getTopEntropyWords(
    guessWords: string[],
    possibleAnswers: string[],
    topN: number = 10
): Array<{ word: string; entropy: number; bitsOfInfo: number }> {
    const allEntropies = calculateAllEntropies(guessWords, possibleAnswers);
    return allEntropies.slice(0, topN);
}

/**
 * Calculates expected remaining possibilities after making a guess
 * This is an alternative metric to entropy - lower is better
 */
export function calculateExpectedRemainingWords(guessWord: string, possibleAnswers: string[]): number {
    if (possibleAnswers.length <= 1) {
        return 0;
    }

    const patternCounts: { [pattern: string]: number } = {};

    for (const answer of possibleAnswers) {
        const pattern = getPattern(guessWord, answer);
        patternCounts[pattern] = (patternCounts[pattern] || 0) + 1;
    }

    // Calculate weighted average of remaining words
    let weightedSum = 0;
    const totalAnswers = possibleAnswers.length;

    for (const count of Object.values(patternCounts)) {
        const probability = count / totalAnswers;
        weightedSum += probability * count;
    }

    return weightedSum;
}

/**
 * Provides a human-readable explanation of entropy
 */
export function explainEntropy(entropy: number): string {
    if (entropy === 0) {
        return "No information gain - only one possibility remains";
    } else if (entropy < 1) {
        return "Low information gain - eliminates few possibilities";
    } else if (entropy < 2) {
        return "Moderate information gain - roughly halves possibilities";
    } else if (entropy < 3) {
        return "Good information gain - reduces possibilities significantly";
    } else if (entropy < 4) {
        return "High information gain - eliminates many possibilities";
    } else {
        return "Excellent information gain - maximum uncertainty reduction";
    }
}

/**
 * Suggests the best starting words based on entropy
 * These are commonly known optimal starting words
 */
export function getBestStartingWords(): Array<{ word: string; description: string }> {
    return [
        { word: "SOARE", description: "Highest entropy starting word (5.89 bits)" },
        { word: "SLATE", description: "Popular high-entropy choice (5.84 bits)" },
        { word: "TRACE", description: "Excellent common-word choice (5.81 bits)" },
        { word: "CRATE", description: "High entropy with common letters (5.78 bits)" },
        { word: "ADIEU", description: "Vowel-heavy strategy (5.76 bits)" },
        { word: "AUDIO", description: "Another vowel-focused option (5.74 bits)" },
        { word: "RAISE", description: "Balanced consonant-vowel mix (5.72 bits)" }
    ];
}

File: src/entropy-wasm/entropy.cpp
#include <emscripten/bind.h>
#include <emscripten/val.h>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
#include <cmath>
#include <iostream>

using namespace emscripten;

class EntropyCalculator {
private:
    std::vector<std::string> allWords;
    std::vector<std::string> possibleAnswers;
    
    // High-performance pattern generation
    std::string getPattern(const std::string& guess, const std::string& target) {
        std::string result(guess.length(), 'B'); // Initialize as all black
        std::vector<int> targetFreq(26, 0);
        
        // Count target letter frequencies
        for (char c : target) {
            targetFreq[c - 'A']++;
        }
        
        // First pass: mark green letters and reduce frequency
        for (size_t i = 0; i < guess.length(); i++) {
            if (guess[i] == target[i]) {
                result[i] = 'G';
                targetFreq[guess[i] - 'A']--;
            }
        }
        
        // Second pass: mark yellow letters
        for (size_t i = 0; i < guess.length(); i++) {
            if (result[i] == 'B' && targetFreq[guess[i] - 'A'] > 0) {
                result[i] = 'Y';
                targetFreq[guess[i] - 'A']--;
            }
        }
        
        return result;
    }

public:
    EntropyCalculator() {}
    
    // Set word lists for calculations
    void setWordLists(const val& allWordsJS, const val& possibleAnswersJS) {
        allWords.clear();
        possibleAnswers.clear();
        
        // Convert JavaScript arrays to C++ vectors
        for (int i = 0; i < allWordsJS["length"].as<int>(); i++) {
            std::string word = allWordsJS[i].as<std::string>();
            std::transform(word.begin(), word.end(), word.begin(), ::toupper);
            allWords.push_back(word);
        }
        
        for (int i = 0; i < possibleAnswersJS["length"].as<int>(); i++) {
            std::string word = possibleAnswersJS[i].as<std::string>();
            std::transform(word.begin(), word.end(), word.begin(), ::toupper);
            possibleAnswers.push_back(word);
        }
        
        std::cout << "🔧 C++ EntropyCalculator initialized with " 
                  << allWords.size() << " total words and " 
                  << possibleAnswers.size() << " possible answers" << std::endl;
    }
    
    // High-performance entropy calculation
    double calculateEntropy(const std::string& guessWord) {
        if (possibleAnswers.size() <= 1) {
            return 0.0;
        }
        
        std::string upperGuess = guessWord;
        std::transform(upperGuess.begin(), upperGuess.end(), upperGuess.begin(), ::toupper);
        
        // Count pattern frequencies using unordered_map for O(1) lookups
        std::unordered_map<std::string, int> patternCounts;
        
        for (const std::string& answer : possibleAnswers) {
            std::string pattern = getPattern(upperGuess, answer);
            patternCounts[pattern]++;
        }
        
        // Calculate Shannon entropy: H = -Σ p(x) * log₂(p(x))
        double entropy = 0.0;
        double totalAnswers = static_cast<double>(possibleAnswers.size());
        
        for (const auto& pair : patternCounts) {
            double probability = pair.second / totalAnswers;
            entropy -= probability * std::log2(probability);
        }
        
        return entropy;
    }
    
    // Ultra-fast bulk entropy calculation for all words
    val calculateAllEntropies() {
        val results = val::array();
        
        if (possibleAnswers.empty()) {
            return results;
        }
        
        std::cout << "🚀 Starting C++ bulk entropy calculation for " 
                  << allWords.size() << " words..." << std::endl;
        
        // Calculate entropy for each word and create result objects
        std::vector<std::pair<double, std::string>> entropyPairs;
        entropyPairs.reserve(allWords.size());
        
        for (const std::string& word : allWords) {
            double entropy = calculateEntropy(word);
            entropyPairs.emplace_back(entropy, word);
        }
        
        // Sort by entropy (highest first) - much faster than JS sorting
        std::sort(entropyPairs.begin(), entropyPairs.end(), 
                  [](const auto& a, const auto& b) { return a.first > b.first; });
        
        // Convert to JavaScript format
        for (const auto& pair : entropyPairs) {
            val result = val::object();
            result.set("word", pair.second);
            result.set("entropy", pair.first);
            result.set("bitsOfInfo", std::round(pair.first * 100.0) / 100.0);
            results.call<void>("push", result);
        }
        
        std::cout << "✅ C++ bulk entropy calculation completed!" << std::endl;
        return results;
    }
    
    // Fast word filtering with constraints
    val filterWords(const val& wordsJS, 
                   const val& knownPositionsJS,
                   const val& yellowLettersJS, 
                   const val& grayLettersJS) {
        
        std::vector<std::string> words;
        std::vector<std::string> knownPositions;
        std::vector<std::string> grayLetters;
        
        // Convert JavaScript arrays to C++ vectors
        for (int i = 0; i < wordsJS["length"].as<int>(); i++) {
            std::string word = wordsJS[i].as<std::string>();
            std::transform(word.begin(), word.end(), word.begin(), ::toupper);
            words.push_back(word);
        }
        
        for (int i = 0; i < knownPositionsJS["length"].as<int>(); i++) {
            std::string pos = knownPositionsJS[i].as<std::string>();
            std::transform(pos.begin(), pos.end(), pos.begin(), ::toupper);
            knownPositions.push_back(pos);
        }
        
        for (int i = 0; i < grayLettersJS["length"].as<int>(); i++) {
            std::string letter = grayLettersJS[i].as<std::string>();
            std::transform(letter.begin(), letter.end(), letter.begin(), ::toupper);
            grayLetters.push_back(letter);
        }
        
        // Process yellow letters (more complex structure)
        std::vector<std::pair<char, std::vector<int>>> yellowLetters;
        for (int i = 0; i < yellowLettersJS["length"].as<int>(); i++) {
            val yellowItem = yellowLettersJS[i];
            char letter = yellowItem["letter"].as<std::string>()[0];
            letter = std::toupper(letter);
            
            std::vector<int> excludedPositions;
            val positions = yellowItem["excludedPositions"];
            for (int j = 0; j < positions["length"].as<int>(); j++) {
                excludedPositions.push_back(positions[j].as<int>());
            }
            yellowLetters.emplace_back(letter, excludedPositions);
        }
        
        // High-performance filtering
        std::vector<std::string> filtered;
        filtered.reserve(words.size() / 4); // Reserve space for efficiency
        
        for (const std::string& word : words) {
            bool valid = true;
            
            // Check known positions (green letters)
            for (size_t i = 0; i < knownPositions.size() && i < word.length(); i++) {
                if (!knownPositions[i].empty() && word[i] != knownPositions[i][0]) {
                    valid = false;
                    break;
                }
            }
            
            if (!valid) continue;
            
            // Check yellow letters
            for (const auto& yellow : yellowLetters) {
                char letter = yellow.first;
                const auto& excludedPositions = yellow.second;
                
                // Word must contain the letter
                if (word.find(letter) == std::string::npos) {
                    valid = false;
                    break;
                }
                
                // Letter must not be in excluded positions
                for (int pos : excludedPositions) {
                    if (pos < word.length() && word[pos] == letter) {
                        valid = false;
                        break;
                    }
                }
                if (!valid) break;
            }
            
            if (!valid) continue;
            
            // Check gray letters
            for (const std::string& grayLetter : grayLetters) {
                if (!grayLetter.empty() && word.find(grayLetter[0]) != std::string::npos) {
                    valid = false;
                    break;
                }
            }
            
            if (valid) {
                filtered.push_back(word);
            }
        }
        
        // Convert back to JavaScript array
        val result = val::array();
        for (const std::string& word : filtered) {
            result.call<void>("push", word);
        }
        
        return result;
    }
};

// Bind the class to JavaScript
EMSCRIPTEN_BINDINGS(entropy_calculator) {
    class_<EntropyCalculator>("EntropyCalculator")
        .constructor<>()
        .function("setWordLists", &EntropyCalculator::setWordLists)
        .function("calculateEntropy", &EntropyCalculator::calculateEntropy)
        .function("calculateAllEntropies", &EntropyCalculator::calculateAllEntropies)
        .function("filterWords", &EntropyCalculator::filterWords);
}